src/ngx_http_lua_kong_module.c

```diff
@@ -670,5 +683,247 @@ ngx_http_lua_kong_get_upstream_ssl_verify(ngx_http_request_t *r,
 
     return ctx->upstream_ssl_verify;
 }
+# endif
+
+// char *
+// ngx_http_lua_kong_load_var_index(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+// {
+
+//     ngx_str_t                     *value;
+//     ngx_int_t                      index;
+
+//     value = cf->args->elts;
+
+//     if (value[1].len == 0) {
+//         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+//                            "invalid variable name size \"%V\"",
+//                            &value[1]);
+//         return NGX_CONF_ERROR;
+//     }
+
+//     index = ngx_http_get_variable_index(cf, &value[1]);
+
+//     if (index == NGX_ERROR) {
+//         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+//                            "variable \"%V\" is not found",
+//                            &value[1]);
+//         return NGX_CONF_ERROR;
+//     }
+
+//     return NGX_CONF_OK;
+// }
+
+int
+ngx_http_lua_kong_ffi_var_load_index(u_char *name_data,
+    size_t name_len, ngx_uint_t *index, char **err)
+{
+    ngx_uint_t                  i;
+    ngx_http_variable_t        *v;
+    ngx_str_t                   name;
+    ngx_http_core_main_conf_t  *cmcf;
+
+    name.data = name_data;
+    name.len = name_len;
+
+    cmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle, ngx_http_core_module);
+
+    /* From src/http/ngx_http_variables.c:ngx_http_get_variable_index */
+    v = cmcf->variables.elts;
+
+    if (v == NULL) {
+        /* This function will always be called post config, so we can assume the variable
+            array always exists
+        */
+        *err = "can't access variables array";
+        return NGX_ERROR;
+
+    } else {
+        for (i = 0; i < cmcf->variables.nelts; i++) {
+            if (name.len != v[i].name.len
+                || ngx_strncasecmp(name.data, v[i].name.data, name.len) != 0)
+            {
+                continue;
+            }
+
+            *index = i;
+            return NGX_OK;
+        }
+    }
+
+    v = ngx_array_push(&cmcf->variables);
```

Comment:
Yeah I agree it might be a bad idea if we allow user to call this after config phase, if it's during request then on L760
we may end up allocating the variable in the request memory pool (correct me if i'm wrong) and that's bad.
How do you feel if we restrict the API to has to be called in `init_by_lua` as it's technically still in config phase?
The reason I didn't manage to use `ngx_http_get_variable_index` here is it takes a `ngx_conf_t` as parameter and I think
that make it only possible to be called inside a directive 🤔